File: app/src/main/java/com/example/securepool/ui/poolgame/Ball.kt
package com.example.localpoolsimulator

import android.graphics.Paint
import com.example.securepool.ui.poolgame.GameConstants
import kotlin.math.abs

/**
 * Represents a single ball in the pool game.
 * Now uses constants from the GameConstants object for physics properties.
 *
 * @param id A unique identifier for the ball (e.g., "cue", "8-ball").
 * @param x Current X coordinate of the ball's center.
 * @param y Current Y coordinate of the ball's center.
 * @param radius Radius of the ball.
 * @param color Color of the ball.
 */
data class Ball(
    val id: String,
    var x: Float,
    var y: Float,
    val radius: Float,
    val color: Int,
) {
    var velocityX: Float = 0f
    var velocityY: Float = 0f

    val paint: Paint = Paint().apply {
        this.color = color
        isAntiAlias = true
    }

    /**
     * Updates the ball's position based on its velocity and applies friction.
     */
    fun update() {
        // Apply friction (a simple deceleration)
        velocityX *= GameConstants.FRICTION_FACTOR
        velocityY *= GameConstants.FRICTION_FACTOR

        // Stop if velocity is very low to prevent infinite crawling
        if (abs(velocityX) < GameConstants.STOP_VELOCITY_THRESHOLD) velocityX = 0f
        if (abs(velocityY) < GameConstants.STOP_VELOCITY_THRESHOLD) velocityY = 0f

        // Update position using the fixed time step
        x += velocityX * GameConstants.PHYSICS_TIME_STEP_SEC
        y += velocityY * GameConstants.PHYSICS_TIME_STEP_SEC
    }

    /**
     * Checks if the ball is currently moving.
     * @return True if the ball has significant velocity, false otherwise.
     */
    fun isMoving(): Boolean {
        return velocityX != 0f || velocityY != 0f
    }
}

File: backend/ball.js
/**
 * @file Ball.js
 * Represents a single ball in the pool game for the server-side simulation.
 */

// This threshold determines when a ball is considered "stopped".
const STOP_THRESHOLD = 1.0;
const BALL_RADIUS = 22.5;

export class Ball {
    /**
     * @param {number} x Initial X coordinate.
     * @param {number} y Initial Y coordinate.
     * @param {string} id A unique identifier for the ball (e.g., 'cue', '8-ball').
     */
    constructor(x, y, id) {
        this.x = x;
        this.y = y;
        this.radius = BALL_RADIUS;
        this.id = id; // Useful for identifying balls on the client

        this.velocityX = 0;
        this.velocityY = 0;
    }

    /**
     * Updates the ball's position based on its velocity and applies friction.
     * This logic should mirror the Kotlin implementation.
     * @param {number} deltaTime Time elapsed since the last update (in seconds).
     */
    update(deltaTime) {
        // Apply friction (a simple deceleration)
        const frictionFactor = 0.99;
        this.velocityX *= frictionFactor;
        this.velocityY *= frictionFactor;

        // Stop if velocity is very low to prevent infinite crawling
        if (Math.abs(this.velocityX) < STOP_THRESHOLD) {
            this.velocityX = 0;
        }
        if (Math.abs(this.velocityY) < STOP_THRESHOLD) {
            this.velocityY = 0;
        }

        // Update position
        this.x += this.velocityX * deltaTime;
        this.y += this.velocityY * deltaTime;
    }

    /**
     * Checks if the ball is currently moving.
     * @returns {boolean} True if the ball has any significant velocity.
     */
    isMoving() {
        return this.velocityX !== 0 || this.velocityY !== 0;
    }
}